libkatana_python_native/src/PropertyGraph.cpp

```diff
@@ -0,0 +1,748 @@
+          KATANA_CHECKED(self.MakeNodeIndex(name));
+        }
+        return self.GetNodeIndex(name);
+      },
+      py::arg("name"), py::return_value_policy::reference_internal);
+  cls.def("has_edge_index", &PropertyGraph::HasEdgeIndex, py::arg("name"));
+  cls.def(
+      "get_edge_index",
+      [](PropertyGraph& self, const std::string& name)
+          -> Result<std::shared_ptr<
+              katana::EntityIndex<katana::GraphTopology::Edge>>> {
+        if (!self.HasEdgeIndex(name)) {
+          KATANA_CHECKED(self.MakeEdgeIndex(name));
+        }
+        return self.GetEdgeIndex(name);
+      },
+      py::arg("name"), py::return_value_policy::reference_internal);
+
+  cls.def("unload_topologies", &PropertyGraph::DropAllTopologies);
+
+  cls.def(
+      "write",
+      [](PropertyGraph& self, const std::string* path,
+         const std::string& provenance) {
+        if (path) {
+          return self.Write(*path, provenance);
+        }
+        return self.Commit(provenance);
+      },
+      py::arg("path") = py::none(),
+      py::arg("provenance") = py::str("katana.local"),
+      py::call_guard<py::gil_scoped_release>(),
+      R"""(
+      Write the property graph to the specified path or URL (or the original path it was loaded from if path is
+      not provided). Provide lineage information in the form of a command line.
+
+      :param path: The path to which to write or None to use `self.path`.
+      :type path: str or Path
+      :param command_line: Lineage information in the form of a command line.
+      :type command_line: str
+      )""");
+  cls.def_property_readonly("path", &PropertyGraph::rdg_dir);
+}
+
+template <typename node_or_edge>
+struct WrapPrimitiveEntityIndex {
+  py::class_<katana::EntityIndex<node_or_edge>> base_cls;
+
+  template <typename T>
+  py::object instantiate(py::module& m, const char* name) {
+    using Cls = katana::PrimitiveEntityIndex<node_or_edge, T>;
+    py::class_<Cls, std::shared_ptr<Cls>> cls(m, name, base_cls);
+
+    cls.template def(
+        "__getitem__", [](Cls& self, const T& v) { return *self.Find(v); });
+    cls.template def("find_all", [](Cls& self, const T& v) {
+      return py::make_iterator(self.LowerBound(v), self.UpperBound(v));
+    });
+
+    //    katana::DefConventions(cls);
+
+    return std::move(cls);
+  }
+};
+
+template <typename node_or_edge>
+struct WrapStringEntityIndex {
+  py::class_<katana::EntityIndex<node_or_edge>> base_cls;
+
+  py::object instantiate(py::module& m, const char* name) {
+    using Cls = katana::StringEntityIndex<node_or_edge>;
+    py::class_<Cls, std::shared_ptr<Cls>> cls(m, name, base_cls);
+
+    cls.template def("__getitem__", [](Cls& self, const std::string& v) {
+      return *self.Find(v);
+    });
+    cls.template def("find_all", [](Cls& self, const std::string& v) {
+      return py::make_iterator(self.LowerBound(v), self.UpperBound(v));
+    });
+
+    //    katana::DefConventions(cls);
+
+    return std::move(cls);
```

Comment:
Does this have to be wrapped in `std::move`?

Reply:
It doesn't. I removed the move.

Reply:
I was wrong. It is needed. For some reason I get a "copied despite being returned by name" warning. Doing the move is harmless, so I'll just leave it there.